### PR TITLE
[DependencyInjection] Add `#[Autoconfigure]` to help define autoconfiguration rules

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Being inspired by the discussion with @derrabus in #39776.

This PR allows declaring autoconfiguration rules using an attribute on classes/interfaces, eg:
`#[Autoconfigure(bind: ['$foo' => 'bar'], tags: [...], calls: [...])]`

This should typically be added on a base class/interface to tell *how* implementations of such a base type should be autoconfigured. The attribute is parsed when autoconfiguration is enabled, except when a definition has the `container.ignore_attributes` tag, which allows opting out from this behavior.

As usual, the corresponding rules are applied only to services that have autoconfiguration enabled.

In practice, this means that this enables auto-tagging of all implementations of this interface:
```php
#[Autoconfigure(tags: ['my_tag'])]
interface MyInterface {...}
```

Of course, all auto-configurable settings are handled (calls, bindings, etc.)

This PR adds another attribute: `#[AutoconfigureTag()]`.

It extends `#[Autoconfigure]` and allows for specifically defining tags to attach by autoconfiguration.

The name of the tag is optional and defaults to the name of the tagged type (typically the FQCN of an interface). This should ease with writing locators/iterators of tagged services.

```php
#[AutoconfigureTag()]
interface MyInterface {...}
```

### CODE REVIEW

File: src/Symfony/Contracts/Service/Attribute/Autoconfigure.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Symfony\Contracts\Service\Attribute;
+
+/**
+ * A tag to tell how a base type should be autoconfigured.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+#[\Attribute(\Attribute::TARGET_CLASS | \Attribute::IS_REPEATABLE)]
+class Autoconfigure
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/Service/Attribute/Autoconfigure.php]: The attribute is not `final` on purpose: subclassing it would allow implementing `AutoTag\*` attributes as discussed in https://github.com/symfony/symfony/pull/39776#issuecomment-757836140

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 2]
```diff
@@ -167,6 +168,13 @@ private function findClasses(string $namespace, string $pattern, array $excludeP
             }
         }
 
+        if (80000 <= \PHP_VERSION_ID) {
+            $parseDefinition = new \ReflectionMethod(YamlFileLoader::class, 'parseDefinition');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: Reusing the implementation in `YamlFileLoader` provides compat with nice syntaxes allowed by the loader, including references.
This won't provide support for tagged arguments, but I don't think we need to provide support for them, for now at least.

File: src/Symfony/Contracts/Service/Attribute/Autoconfigure.php (hunk 1 above)
Comment (ro0NL) [reply src/Symfony/Contracts/Service/Attribute/Autoconfigure.php]: i tend to be skeptical about sub-classing this attribute in general :)

mostly in terms of the extra keyword(s) that then need to be accounted for (eg. search in project) as DI config diverges

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Service/Attribute/Autoconfigure.php]: I think there is no reason to close this. I also like the DX that would be provided by the ability to type `AutoTag\` and be suggested with all existing tags, no matter the leading namespace.

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 3]
```diff
@@ -96,7 +97,7 @@ public function registerClasses(Definition $prototype, string $namespace, string
             throw new InvalidArgumentException(sprintf('Namespace is not a valid PSR-4 prefix: "%s".', $namespace));
         }
 
-        $classes = $this->findClasses($namespace, $resource, (array) $exclude);
+        $classes = $this->findClasses($namespace, $resource, (array) $exclude, $prototype->isAutoconfigured());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: The attribute is parsed only if the autodiscovery group has autoconfiguration enabled.
This extra check could be removed. Let me know how you feel about this.

File: src/Symfony/Contracts/Service/Attribute/Autoconfigure.php (hunk 1 above)
Comment (ro0NL) [reply src/Symfony/Contracts/Service/Attribute/Autoconfigure.php]: i've changed my mind. All this is an implementation detail of `autoconfigure: true` from a consumer pov, and it should just work :)

the rest is then solved by https://github.com/symfony/symfony/issues/26295, which becomes more relevant btw ;)

**Comment by ro0NL** [comment]
is there any reason for the "auto discovered services only" actually?

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php [hunk 4]
```diff
@@ -207,6 +215,17 @@ private function findClasses(string $namespace, string $pattern, array $excludeP
             if ($r->isInstantiable() || $r->isInterface()) {
                 $classes[$class] = null;
             }
+
+            if (80000 > \PHP_VERSION_ID || !$autoconfigure) {
+                continue;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: (@ro0NL using threads to allow scoped discussions)

> is there any reason for the "auto discovered services only" actually?

Yes: the attribute must be parsed before compiler passes, and when classes are loaded. The DI container of Symfony doesn't load class definitions if it doesn't have to. When autodiscovery is enabled, we know we need to load the corresponding classes, so that's the proper hook.

Also because it just makes sense - autodiscovery means that :)

Comment (ro0NL) [reply src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: given

```
_defaults { autoconfigure: true }

Foo\: { prototype: 'path/to/foos' }
Bar\Baz: ~
```

conceptually i expect the attribute to take effect when it's defined on either `Bar\Baz` or `Foo\*`, and because of `autoconfigure: true`

when it's defined on abstracts/interfaces i expect it to behave the same as `_instanceof` directives would.

Comment (ro0NL) [reply src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: alternatively, what seems to be the route you're taking now: `#[ServicePrototype(...)]`

or `#[ServicePrototypeAutoconfig(...)]` to go fully explicit in terms of exposing this is enabled thru `autoconfigure: true` for prototyped services.

File: src/Symfony/Component/DependencyInjection/Loader/FileLoader.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Loader/FileLoader.php]: Closing as I think this is the correct behavior.

**Comment by apfelbox** [comment]
Quick nitpick question: should the attribute be named `#[Autoconfigure]` or `#[Autoconfigured]`? Because the class is autoconfigured, it does not do any configuring itself.

The recently proposed `#[Deprecated]` attribute in PHP itself is in passive voice as well.

**Comment by nicolas-grekas** [comment]
> Because the class is autoconfigured, it does not do any configuring itself.

I feel like I missed explaining correctly how this works. *Yes*, this attribute declares a configuration rule for autoconfiguration.

File: src/Symfony/Component/Console/Attribute/CommandAutoTag.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+#[\Attribute(\Attribute::TARGET_CLASS | \Attribute::IS_REPEATABLE)]
+class CommandAutoTag extends AutoTag
+{
+    public const NAME = 'console.command';
+
+    public function __construct(string $name)
+    {
+        parent::__construct(tags: [[
```
Comment (OskarStark) [inline src/Symfony/Component/Console/Attribute/CommandAutoTag.php]: Based on Travis, this line produces a syntax error on PHP 7.2 🤔

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.3.0
+-----
+
+ * added support for loading autoconfiguration rules via the `#[Autoconfigure]` attribute on PHP 8
```
Comment (OskarStark) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]:
```suggestion
5.3
---

 * Add support for loading autoconfiguration rules via the `#[Autoconfigure]` attribute on PHP 8
```

File: src/Symfony/Component/EventDispatcher/CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.3.0
+-----
+
+ * Added `Attribute\EventListenerAutoTag` to help autoconfigure event listeners
```
Comment (OskarStark) [inline src/Symfony/Component/EventDispatcher/CHANGELOG.md]:
```suggestion
5.3
---

 * Add `Attribute\EventListenerAutoTag` to help autoconfigure event listeners
```

File: src/Symfony/Contracts/CHANGELOG.md [hunk 8]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+2.4.0
+-----
+
+ * added `Service\Attribute\Auto*` to help define autoconfiguration rules for PHP 8
+
```
Comment (OskarStark) [inline src/Symfony/Contracts/CHANGELOG.md]:
```suggestion
2.4
---

 * Add `Service\Attribute\Auto*` to help define autoconfiguration rules for PHP 8

```

File: src/Symfony/Component/Console/Attribute/CommandAutoTag.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Attribute/CommandAutoTag.php]: outdated CI, now rebased

**Comment by nicolas-grekas** [comment]
PR updated: registration happens in a new compiler pass and doesn't need autodiscovery anymore.
I also added a new `container.ignore_attributes` tag to disable processing `Autoconfigure` attributes when needed, for more flexibility.

**Comment by nicolas-grekas** [comment]
In practice, this PR enables tagging some service classes, eg.:
```php
#[EventListenerAutoTag(event: 'kernel.request')]
class MyRequestListener {...}
```

but also defining autoconfiguration rules for some interfaces, eg.:
```php
#[AutoCall(method: 'setLogger', arguments: ['@my_logger'])]
interface MyLoggerAwareInterface {...}
```

These examples use child classes of the base `Autoconfigure`, which enable better DX, but are basically just a specialization of it.

**Comment by nicolas-grekas** [comment]
This PR also adds specialized attributes for declaring individual autoconfiguration rules:

- `#[AutoBind(['string $apiKey' => '%some_api_key%'])]`
- `#[AutoCall(method: 'setLogger', arguments: [...]))`
- `#[AutoTag(name: 'my_tag', attributes => [...])]

Of course, it's also possible to create specialized variants of those specializations.
This PR embed two of them, for the example:
- `#[EventListenerAutoTag(event: 'my_event')]`
- `#[CommandAutoTag(name: 'app:my_command')]` (should allow replacing the `$defaultName` property once #39851 is merged)

All attributes are defined at the class level. I did not try to define them at the method/argument level - that'd be totally doable in the future PR.

File: src/Symfony/Contracts/Service/Attribute/Autoconfigure.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\Service\Attribute;
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/Service/Attribute/Autoconfigure.php]: I've added the `Autoconfigure` attribute in contracts because I believe it conveys semantics that could benefit any DI container. But I'm fine moving it to the DI component if that's preferred.

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Service/Attribute/Autoconfigure.php]: (but this would prevent using the attribute without the DI component - while e.g. the Console component would benefit from the semantics to define commands' name + description + etc)

**Comment by nicolas-grekas** [comment]
PR is now tested, green and ready! (fabbot's failure is a false positive)

Be careful, it's not anymore a replacement for #39776, but more like a sidekick of #39897.

Unlike the previous version, there is nothing in contracts anymore, but a single `Autoconfigure` attribute in the DI component.

This attribute is meant to target one use case: defining autoconfiguration rules bound to interfaces, as in e.g.:
```php
#[Autoconfigure(tags: ['my_tag'])]
interface MyInterface {...}
```

Of course, all auto-configurable settings are handled (calls, bindings, etc.)

/cc @derrabus, vote pending ;)

File: src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
+
+/**
+ * Reads #[Autoconfigure] attributes on definitions that are autoconfigured
+ * and don't have the "container.ignore_attributes" tag.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
```
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php:24]: `@final` to let us some flexibility on the public api?

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php (hunk 10 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php:24]: updated

**Comment by nicolas-grekas** [comment]
I pushed a new idea: a new `#[AutoconfigureTag()]` attribute.

It extends `#[Autoconfigure]` but allows for specifically defining tags to attach by autoconfiguration.

The name of the tag is optional and defaults to the name of the tagged type (typically the FQCN of an interface). This should ease with writing locators/iterators of tagged services.

```php
#[AutoconfigureTag()]
interface MyInterface {...}
```

File: src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+final class RegisterAutoconfigureAttributesPass implements CompilerPassInterface
+{
+    private $ignoreAttributesTag;
+    private $registerForAutoconfiguration;
+
+    public function __construct(string $ignoreAttributesTag = 'container.ignore_attributes')
```
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php:30]: should probably be ignored in `UnusedTagsPass`

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 12]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 ---
 
  * Add `ServicesConfigurator::remove()` in the PHP-DSL
+ * Add support for loading autoconfiguration rules via the `#[Autoconfigure]` attribute on PHP 8
```
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]: ``` and `#[AutoconfigureTag]` attributes```

**Review by chalasr — APPROVED** [review]

**Comment by wouterj** [comment]
IIUC, we now have 2 competing PRs here. As far as I know, it was decided that progress continues in #39897? Please correct me if I'm wrong, but otherwise this one should be closed I think.

**Comment by nicolas-grekas** [comment]
@wouterj this one provides something else. Both should be merged to me in the end. This one before #39897 since it's ready and might be used by #39897 as explained in https://github.com/symfony/symfony/pull/39897#discussion_r561029679

File: src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php (hunk 11 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php:30]: we don't use it anywhere, so no need to

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md (hunk 12 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/CHANGELOG.md]: updated.

File: src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php (hunk 10 above)
Comment (jderusse) [reply src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php:24]: you changed your mind?

**Review by jderusse — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php (hunk 10 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php:24]: I made the class final

**Comment by wouterj** [comment]
As requested, I'll post some of my comment in the other PR (ref: https://github.com/symfony/symfony/pull/39897#discussion_r571675897 ) here:

:smiley: What I like about this PR is that it reuses the autoconfigure system (using the Yaml file loader seems a bit hacky, but I'm surely not the one to make a decision about that).
:neutral_face: What I'm not so sure about is why there is a need to have a userland `Autoconfigure` attribute. I personally don't see this need (especially if #39897 will make it, as that implements more specialized and nice userland attributes for this purpose).
I think such a generic attribute opens up a lot of misuse (e.g. in the other PR, it was shown that you can misuse attributes instead of implementing the `ResetInterface` contract).  
I feel like it couples attributes directly to DI (which e.g. a `@required` annotation explicitly avoids).  
At last, by having both a generic attribute and more specific attributes (from #39897), we provide more than 1 solution for the same problem. How should we document this? What are the advantages of one over the other?

**Comment by nicolas-grekas** [comment]
> (using the Yaml file loader seems a bit hacky, but I'm surely not the one to make a decision about that).

What matters here is how one can describe something using an array data structure. YamlLoader does exactly this: turn arrays (from yaml) into actual calls. Here, we just reuse that logic, so that the same array-structures can also be used in attributes. E.g.:
```
#[Autoconfigure(
    calls: [
		'setFoo' => ['the-args']
	]
)]
```

> What I'm not so sure about is why there is a need to have a userland Autoconfigure attribute. I personally don't see this need

All ppl that use `_instanceof` or `$container->registerForAutoconfiguration()` have a need for this.
It allows telling to the DI engine how a definition should be autoconfigured, depending on the base types of the class it describes.

> especially if #39897 will make it, as that implements more specialized and nice userland attributes for this purpose.

There is a critical difference with #39897: the attribute in this PR allows defining rules that are *inherited*. #39897 defines attributes that are *not* inherited.

That's *THE* critical difference between both PRs, and why they fill a different need.

> I think such a generic attribute opens up a lot of misuse (e.g. in the other PR, it was shown that you can misuse attributes instead of implementing the ResetInterface contract).

Can you clarify? To me, it looks like you're arguing against empowering ppl, which I'm sure you aren't.
I must miss something, or I just don't agree :)

> I feel like it couples attributes directly to DI (which e.g. a @required annotation explicitly avoids).

I do not understand this "coupling" argument. To me, as I explained in https://github.com/symfony/symfony/pull/39776#issuecomment-762326163, there is zero difference between this PR, #39776 or #39897 in this regard.

Maybe you can help me understand what you mean with actual examples of when this "coupling" would happen?

> At last, by having both a generic attribute and more specific attributes (from #39897), we provide more than 1 solution for the same problem. How should we document this? What are the advantages of one over the other?

See "the critical difference" I mention above.

**Comment by nicolas-grekas** [comment]
> How should we document this?

TL;DR of my previous comment:
- use `#[Autoconfigure]` on interfaces
- use #39897 on specific class-services

Note that these two lines are orthogonal.

**Review by derrabus — APPROVED** [review]
I understand the intention of this generic attribute as a nicer way to configure autoconfiguration rules for userland base types.

Instead of cluttering my kernel with…

```php
$container->registerForAutoconfiguration(MyInterface::class) 
    ->addTag('my_tag');
```

… I annotate this rule to my interface instead.

```php
#[Autoconfigure(tags: ['my_tag'])]
interface MyInterface {}
```

Did I get this right?

I think, this is the use case we can agree on. But since attributes are a new feature to most users, it is imperative that we distinguish the use-cases of this PR and #39897 in the documentation.

**Comment by nicolas-grekas** [comment]
@derrabus you've got it right. I agree about the doc.

**Comment by kozlice** [comment]
This is a really cool feature, but I've just tried it and there seems to be an issue:

```php
#[Autoconfigure(tags: ['some.tag'])]
interface MyServiceInterface {}

class MyService implements MyServiceInterface {}

$class = new ReflectionClass(MyService::class);
// That's how RegisterAutoconfigureAttributesPass does it
$attributes = $class->getAttributes(Autoconfigure::class, \ReflectionAttribute::IS_INSTANCEOF);
// Whoopsie, empty array
var_dump($attributes);
```

Running PHP 8.0.6. Looks like the `IS_INSTANCEOF` flag doesn't do what it is supposed to do.

**Comment by nicolas-grekas** [comment]
You might have missed how this works. I can't really help because I don't know what you want todo, but of course, requesting the attributes on class `MyService` won't list  you the attributes on interface `MyServiceInterface`...

**Comment by kozlice** [comment]
@nicolas-grekas Yeah, I might've misunderstood something.

[The announcement in blog](https://symfony.com/blog/new-in-symfony-5-3-service-autoconfiguration-and-attributes) says that `Autoconfigure` can be used to replace `registerForAutoconfiguration` for auto-tagging by interface.

So, I've tried this:

```php
#[Autoconfigure(tags: ['some.tag'])]
interface ServiceInterface {}

class ServiceA implements ServiceInterface {}
class ServiceB implements ServiceInterface {}
```

```yaml
services:
    _defaults:
        autowire: true
        autoconfigure: true

    App\ServiceA: ~
    App\ServiceB: ~
```

And I was expecting both services to be tagged, but they are not.

**Comment by nicolas-grekas** [comment]
Please open an issue if you think you found a bug, with a reproducer ideally.
